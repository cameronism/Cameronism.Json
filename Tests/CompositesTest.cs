using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using Xunit;

namespace Tests
{
	public class CompositesTest
	{
		class B<T>
		{
			public T i { get; set; }
		}

		class B<T1, T2>
		{
			public T1 i { get; set; }
			public T2 j { get; set; }
		}

		public struct MyStructWithFields<T1, T2, T3>
		{
			public T1 i;
			public T2 j { get; set; }
			public T3 k;
		}

		static B<T> A<T>(T i)
		{
			return new B<T> { i = i };
		}

		static B<T1, T2> A<T1, T2>(T1 i, T2 j)
		{
			return new B<T1, T2> { i = i, j = j };
		}

		static Type GetType<T>(T sample)
		{
			return typeof(T);
		}

		[Fact]
		public void Reflect()
		{
			int failed = 0;
			var sb = new StringBuilder();

			foreach (var t in new[] { 
				typeof(System.Int32),
				typeof(System.Boolean),
				typeof(System.Byte),
				typeof(System.Char),
				typeof(System.DateTime),
				typeof(System.Decimal),
				typeof(System.Double),
				typeof(System.Int16),
				typeof(System.Int64),
				typeof(System.SByte),
				typeof(System.Single),
				typeof(System.String),
				typeof(System.UInt16),
				typeof(System.UInt32),
				typeof(System.UInt64),

				typeof(System.Int32?),
				typeof(System.Boolean?),
				typeof(System.Byte?),
				typeof(System.Char?),
				typeof(System.DateTime?),
				typeof(System.Decimal?),
				typeof(System.Double?),
				typeof(System.Int16?),
				typeof(System.Int64?),
				typeof(System.SByte?),
				typeof(System.Single?),
				typeof(System.UInt16?),
				typeof(System.UInt32?),
				typeof(System.UInt64?),

				typeof(System.Guid),
				typeof(System.Guid?),

				GetType(A(1)),
				GetType(A(1, 2)),
				GetType(A(1, A(1))),
				GetType(A(A(1), 1)),
				GetType(A(A(1, A(1)), 1)),
				GetType(A(1, A(A(1, A(1))))),
				GetType(A(Guid.Empty, A(A(1, A(1, (Guid?)null))))),
				typeof(KeyValuePair<int, int>),
				typeof(KeyValuePair<int, int>?),
				typeof(MyStructWithFields<int, int?, string>),
				typeof(MyStructWithFields<int, int?, string>?),

				//typeof(string[]),
				//typeof(IEnumerable<int>),
				//typeof(IReadOnlyList<Guid>),

				//GetType(new[] { new { s = "" } }.ToList()),

				//typeof(Dictionary<string, string>),
				//typeof(IReadOnlyDictionary<Guid, System.DayOfWeek>),


				//typeof(DateTimeOffset),
				//typeof(DateTimeOffset?),

				//typeof(System.Net.IPAddress),

				//typeof(ExplicitDataMemberOrder),
			})
			{
				var schema = UnsafeJson.Schema.Reflect(t);
				bool success = (bool)DescribeMethod.MakeGenericMethod(t).Invoke(null, new object[] { sb });
				if (!success) failed++;
			}

			ApprovalTests.Approvals.Verify(sb.ToString());

			Assert.True(failed == 0, "Look at the approval " + failed + " tests failed");
		}

		static readonly MethodInfo DescribeMethod = typeof(CompositesTest).GetMethod("Describe", BindingFlags.Static | BindingFlags.NonPublic);

		static bool Describe<T>(StringBuilder sb)
		{
			sb.AppendLine();
			sb.AppendLine();
			sb.AppendLine("# " + SchemaTest.HumanName(typeof(T)));

			var emit = UnsafeJson.Composites.Create<T>(UnsafeJson.Schema.Reflect(typeof(T)));

			string instructions;
			try
			{
				var writer = emit.CreateDelegate(out instructions);
				sb.AppendLine(instructions);
				return true;
			}
			catch (Sigil.SigilVerificationException sve)
			{
				sb.AppendLine("## failed");
				sb.AppendLine(sve.ToString());

				return false;
			}
		}
	}
}
