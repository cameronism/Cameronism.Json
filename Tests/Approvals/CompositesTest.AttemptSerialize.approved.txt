

# System.Nullable<System.Int32>


## Newtonsoft 0


0000   31                                                  1

## UnsafeJson 0


0000   31                                                  1

### Equal: True

## Newtonsoft 1


0000   32 31 34 37 34 38 33 36   34 37                     21474836 47

## UnsafeJson 1


0000   32 31 34 37 34 38 33 36   34 37                     21474836 47

### Equal: True

## Newtonsoft 2


0000   6e 75 6c 6c                                         null

## UnsafeJson 2


0000   6e 75 6c 6c                                         null

### Equal: True

# System.Nullable<System.Guid>


## Newtonsoft 0


0000   22 30 30 30 30 30 30 30   30 2d 30 30 30 30 2d 30   "0000000 0-0000-0
0010   30 30 30 2d 30 30 30 30   2d 30 30 30 30 30 30 30   000-0000 -0000000
0020   30 30 30 30 30 22                                   00000"

## UnsafeJson 0


0000   22 30 30 30 30 30 30 30   30 2d 30 30 30 30 2d 30   "0000000 0-0000-0
0010   30 30 30 2d 30 30 30 30   2d 30 30 30 30 30 30 30   000-0000 -0000000
0020   30 30 30 30 30 22                                   00000"

### Equal: True

## Newtonsoft 1


0000   22 34 31 64 30 39 31 62   61 2d 64 36 62 62 2d 34   "41d091b a-d6bb-4
0010   37 39 35 2d 39 36 39 61   2d 32 38 64 31 35 30 39   795-969a -28d1509
0020   64 65 36 62 36 22                                   de6b6"

## UnsafeJson 1


0000   22 34 31 64 30 39 31 62   61 2d 64 36 62 62 2d 34   "41d091b a-d6bb-4
0010   37 39 35 2d 39 36 39 61   2d 32 38 64 31 35 30 39   795-969a -28d1509
0020   64 65 36 62 36 22                                   de6b6"

### Equal: True

## Newtonsoft 2


0000   6e 75 6c 6c                                         null

## UnsafeJson 2


0000   6e 75 6c 6c                                         null

### Equal: True

# System.Nullable<System.DayOfWeek>


## Newtonsoft 0


0000   31                                                  1

## UnsafeJson 0


0000   31                                                  1

### Equal: True

## Newtonsoft 1


0000   35                                                  5

## UnsafeJson 1


0000   35                                                  5

### Equal: True

## Newtonsoft 2


0000   6e 75 6c 6c                                         null

## UnsafeJson 2


0000   6e 75 6c 6c                                         null

### Equal: True

# <>f__AnonymousType0


## Newtonsoft 0


0000   7b 7d                                               {}

## UnsafeJson 0


0000   7b 7d                                               {}

### Equal: True

## Newtonsoft 1


0000   6e 75 6c 6c                                         null

## UnsafeJson 1


0000   6e 75 6c 6c                                         null

### Equal: True

# Tests.B<System.Int32>


## Newtonsoft 0


0000   7b 22 69 22 3a 31 7d                                {"i":1}

## UnsafeJson 0


0000   7b 22 69 22 3a 31 7d                                {"i":1}

### Equal: True

## Newtonsoft 1


0000   6e 75 6c 6c                                         null

## UnsafeJson 1


0000   6e 75 6c 6c                                         null

### Equal: True

# Tests.B<System.Int32, System.Int32>


## Newtonsoft 0


0000   7b 22 69 22 3a 31 2c 22   6a 22 3a 32 7d            {"i":1," j":2}

## UnsafeJson 0


0000   7b 22 69 22 3a 31 2c 22   6a 22 3a 32 7d            {"i":1," j":2}

### Equal: True

## Newtonsoft 1


0000   6e 75 6c 6c                                         null

## UnsafeJson 1


0000   6e 75 6c 6c                                         null

### Equal: True

# Tests.B<System.Nullable<System.Int32>, System.Int32>


## Newtonsoft 0


0000   7b 22 69 22 3a 31 2c 22   6a 22 3a 32 7d            {"i":1," j":2}

## UnsafeJson 0


0000   7b 22 69 22 3a 31 2c 22   6a 22 3a 32 7d            {"i":1," j":2}

### Equal: True

## Newtonsoft 1


0000   7b 22 69 22 3a 6e 75 6c   6c 2c 22 6a 22 3a 32 7d   {"i":nul l,"j":2}

## UnsafeJson 1


0000   7b 22 69 22 3a 6e 75 6c   6c 2c 22 6a 22 3a 32 7d   {"i":nul l,"j":2}

### Equal: True

# Tests.B<System.Int32, System.Nullable<System.Int32>>


## Newtonsoft 0


0000   7b 22 69 22 3a 32 2c 22   6a 22 3a 31 7d            {"i":2," j":1}

## UnsafeJson 0


0000   7b 22 69 22 3a 32 2c 22   6a 22 3a 31 7d            {"i":2," j":1}

### Equal: True

## Newtonsoft 1


0000   7b 22 69 22 3a 32 2c 22   6a 22 3a 6e 75 6c 6c 7d   {"i":2," j":null}

## UnsafeJson 1


0000   7b 22 69 22 3a 32 2c 22   6a 22 3a 6e 75 6c 6c 7d   {"i":2," j":null}

### Equal: True

# Tests.B<System.Int32, System.Nullable<System.Int32>>


## Newtonsoft 0


0000   7b 22 69 22 3a 32 2c 22   6a 22 3a 31 7d            {"i":2," j":1}

## UnsafeJson 0


0000   7b 22 69 22 3a 32 2c 22   6a 22 3a 31 7d            {"i":2," j":1}

### Equal: True

## Newtonsoft 1


0000   7b 22 69 22 3a 32 2c 22   6a 22 3a 6e 75 6c 6c 7d   {"i":2," j":null}

## UnsafeJson 1


0000   7b 22 69 22 3a 32 2c 22   6a 22 3a 6e 75 6c 6c 7d   {"i":2," j":null}

### Equal: True

# System.Collections.Generic.KeyValuePair<System.String, System.Guid>


## Newtonsoft 0


0000   7b 22 4b 65 79 22 3a 6e   75 6c 6c 2c 22 56 61 6c   {"Key":n ull,"Val
0010   75 65 22 3a 22 30 30 30   30 30 30 30 30 2d 30 30   ue":"000 00000-00
0020   30 30 2d 30 30 30 30 2d   30 30 30 30 2d 30 30 30   00-0000- 0000-000
0030   30 30 30 30 30 30 30 30   30 22 7d                  00000000 0"}

## UnsafeJson 0


0000   7b 22 4b 65 79 22 3a 6e   75 6c 6c 2c 22 56 61 6c   {"Key":n ull,"Val
0010   75 65 22 3a 22 30 30 30   30 30 30 30 30 2d 30 30   ue":"000 00000-00
0020   30 30 2d 30 30 30 30 2d   30 30 30 30 2d 30 30 30   00-0000- 0000-000
0030   30 30 30 30 30 30 30 30   30 22 7d                  00000000 0"}

### Equal: True

# System.Nullable<System.Collections.Generic.KeyValuePair<System.String, System.Guid>>


## Newtonsoft 0


0000   7b 22 4b 65 79 22 3a 6e   75 6c 6c 2c 22 56 61 6c   {"Key":n ull,"Val
0010   75 65 22 3a 22 30 30 30   30 30 30 30 30 2d 30 30   ue":"000 00000-00
0020   30 30 2d 30 30 30 30 2d   30 30 30 30 2d 30 30 30   00-0000- 0000-000
0030   30 30 30 30 30 30 30 30   30 22 7d                  00000000 0"}

## UnsafeJson 0


0000   7b 22 4b 65 79 22 3a 6e   75 6c 6c 2c 22 56 61 6c   {"Key":n ull,"Val
0010   75 65 22 3a 22 30 30 30   30 30 30 30 30 2d 30 30   ue":"000 00000-00
0020   30 30 2d 30 30 30 30 2d   30 30 30 30 2d 30 30 30   00-0000- 0000-000
0030   30 30 30 30 30 30 30 30   30 22 7d                  00000000 0"}

### Equal: True

## Newtonsoft 1


0000   6e 75 6c 6c                                         null

## UnsafeJson 1


0000   6e 75 6c 6c                                         null

### Equal: True

# Tests.B<Tests.B<System.Int32, System.Int32>, System.Nullable<System.Int32>>


## Newtonsoft 0


0000   7b 22 69 22 3a 7b 22 69   22 3a 33 2c 22 6a 22 3a   {"i":{"i ":3,"j":
0010   34 7d 2c 22 6a 22 3a 31   7d                        4},"j":1 }

## UnsafeJson 0


0000   7b 22 69 22 3a 7b 22 69   22 3a 33 2c 22 6a 22 3a   {"i":{"i ":3,"j":
0010   34 7d 2c 22 6a 22 3a 31   7d                        4},"j":1 }

### Equal: True

## Newtonsoft 1


0000   7b 22 69 22 3a 7b 22 69   22 3a 33 2c 22 6a 22 3a   {"i":{"i ":3,"j":
0010   34 7d 2c 22 6a 22 3a 6e   75 6c 6c 7d               4},"j":n ull}

## UnsafeJson 1


0000   7b 22 69 22 3a 7b 22 69   22 3a 33 2c 22 6a 22 3a   {"i":{"i ":3,"j":
0010   34 7d 2c 22 6a 22 3a 6e   75 6c 6c 7d               4},"j":n ull}

### Equal: True

## Newtonsoft 2


0000   7b 22 69 22 3a 6e 75 6c   6c 2c 22 6a 22 3a 6e 75   {"i":nul l,"j":nu
0010   6c 6c 7d                                            ll}

## UnsafeJson 2


0000   7b 22 69 22 3a 6e 75 6c   6c 2c 22 6a 22 3a 6e 75   {"i":nul l,"j":nu
0010   6c 6c 7d                                            ll}

### Equal: True

# Tests.B<System.Nullable<System.Int32>, Tests.B<System.Int32, System.Int32>>


## Newtonsoft 0


0000   7b 22 69 22 3a 31 2c 22   6a 22 3a 7b 22 69 22 3a   {"i":1," j":{"i":
0010   33 2c 22 6a 22 3a 34 7d   7d                        3,"j":4} }

## UnsafeJson 0


0000   7b 22 69 22 3a 31 2c 22   6a 22 3a 7b 22 69 22 3a   {"i":1," j":{"i":
0010   33 2c 22 6a 22 3a 34 7d   7d                        3,"j":4} }

### Equal: True

## Newtonsoft 1


0000   7b 22 69 22 3a 6e 75 6c   6c 2c 22 6a 22 3a 7b 22   {"i":nul l,"j":{"
0010   69 22 3a 33 2c 22 6a 22   3a 34 7d 7d               i":3,"j" :4}}

## UnsafeJson 1


0000   7b 22 69 22 3a 6e 75 6c   6c 2c 22 6a 22 3a 7b 22   {"i":nul l,"j":{"
0010   69 22 3a 33 2c 22 6a 22   3a 34 7d 7d               i":3,"j" :4}}

### Equal: True

## Newtonsoft 2


0000   7b 22 69 22 3a 6e 75 6c   6c 2c 22 6a 22 3a 6e 75   {"i":nul l,"j":nu
0010   6c 6c 7d                                            ll}

## UnsafeJson 2


0000   7b 22 69 22 3a 6e 75 6c   6c 2c 22 6a 22 3a 6e 75   {"i":nul l,"j":nu
0010   6c 6c 7d                                            ll}

### Equal: True